beetsplug/lyrics.py

```diff
@@ -209,105 +193,151 @@ def slug(text):
     3. strip whitespace
     4. replace other non-word characters with dashes
     5. strip extra dashes
-
-    This somewhat duplicates the :func:`Google.slugify` function but
-    slugify is not as generic as this one, which can be reused
-    elsewhere.
     """
     return re.sub(r"\W+", "-", unidecode(text).lower().strip()).strip("-")
 
 
-if HAS_BEAUTIFUL_SOUP:
+class RequestHandler:
+    _log: beets.logging.Logger
 
-    def try_parse_html(html, **kwargs):
-        return bs4.BeautifulSoup(html, "html.parser", **kwargs)
+    def debug(self, message: str, *args: Any) -> None:
+        """Log a debug message with the class name."""
+        self._log.debug(f"{self.__class__.__name__}: {message}", *args)
 
-else:
+    def info(self, message: str, *args: Any) -> None:
+        """Log an info message with the class name."""
+        self._log.info(f"{self.__class__.__name__}: {message}", *args)
 
-    def try_parse_html(html, **kwargs):
-        return None
+    def warn(self, message: str, *args: Any) -> None:
+        """Log warning with the class name."""
+        self._log.warning(f"{self.__class__.__name__}: {message}", *args)
+
+    def fetch_text(self, url: str, **kwargs) -> str:
+        """Return text / HTML data from the given URL."""
+        self.debug("Fetching HTML from {}", url)
+        return r_session.get(url, **kwargs).text
+
+    def fetch_json(self, url: str, **kwargs):
+        """Return JSON data from the given URL."""
+        self.debug("Fetching JSON from {}", url)
+        return r_session.get(url, **kwargs).json()
+
+    @contextmanager
+    def handle_request(self) -> Iterator[None]:
+        try:
+            yield
+        except requests.JSONDecodeError:
+            self.warn("Could not decode response JSON data")
+        except requests.RequestException as exc:
+            self.warn("Request error: {}", exc)
+
+
+class BackendType(type):
+    """Metaclass for the :class:`Backend` class.
+
+    It keeps track of defined subclasses and provides access to them through
+    the base class:
+
+    >>> Backend["genius"]  # beetsplug.lyrics.Genius
+    >>> list(Backend)  # ["lrclib", "musixmatch", "genius", "tekstowo", "google"]
+    """
 
+    _registry: dict[str, BackendType] = {}
+    REQUIRES_BS: bool
```

Comment:
These should be typed as `ClassVar`s, right?

Reply:
Ah yes, completely!